python/sdk/merlin/model.py

```diff
@@ -1029,82 +1029,67 @@ def deploy(self, environment_name: str = None,
         :param protocol: protocol to be used for deploying the model (default: HTTP_JSON)
         :return: VersionEndpoint object
         """
+        current_endpoint = self.endpoint
+        env_list = self._get_env_list()
 
-        target_env_name = environment_name
-        if target_env_name is None:
-            env_api = EnvironmentApi(self._api_client)
-            env_list = env_api.environments_get()
-            for env in env_list:
-                if env.is_default:
-                    target_env_name = env.name
-
-            if target_env_name is None:
-                raise ValueError("Unable to find default environment, "
-                                 "pass environment_name to the method")
-
-        if resource_request is None:
-            env_api = EnvironmentApi(self._api_client)
-            env_list = env_api.environments_get()
-
-            for env in env_list:
-                if env.name == target_env_name:
-                    resource_request = ResourceRequest(
-                        env.default_resource_request.min_replica,
-                        env.default_resource_request.max_replica,
-                        env.default_resource_request.cpu_request,
-                        env.default_resource_request.memory_request,
-                        )
-
-            # This case is when the default resource request is not specified in the environment config
-            if resource_request is None:
-                raise ValueError("resource request must be specified")
-
-        resource_request.validate()
+        # target_env_name can never be set as null
+        target_env_name = environment_name if environment_name is not None else ModelVersion._get_default_target_env_name(env_list)
```

Comment:
Not sure if this is an intended or not, but it was never possible to make API calls via the SDK with `environment_name` being `None` or `""`. When creating or updating an endpoint, the SDK would always either use the user-given env name, or set a default env name, or raise an error if it can't find one.

The behaviour of this field is also different from the other fields when updating an endpoint - you can't leave the new endpoint config's environment name empty and expect the API server to redeploy the endpoint in the same environment (an error will be returned from the API server instead).

As such, since this field works differently from the other config fields (e.g. deployment mode, protocol, logger, etc.), I'm assigning the value to this variable separately.

Reply:
In CreateEndpoint API, if env_name is empty, we have logic to get default environment ([ref](https://github.com/caraml-dev/merlin/blob/main/api/api/version_endpoints_api.go#L138)). However, for UpdateEndpoint API, the env_name must be specified ([ref](https://github.com/caraml-dev/merlin/blob/main/api/api/version_endpoints_api.go#L262)). So, I think for redeployment, the `target_env_name` must be specified with current_endpoint's env_name if it's None.

Reply:
Otherwise, if a model already deployed to non-default env and got redeployment without setting env_name will be redeployed to the default env. For example:

```
# First deployment will deploy the model version to non-default-env
merlin.deploy(v, environment_name="non-default-env")

# Redeployment without environment_name will target default-env
merlin.deploy(v)
```

Reply:
Sounds good, I'll make `target_env_name` be set to the current endpoint's `env_name` if it's None. Thanks for the really clear explanation! :D

Reply:
Okay sorry to bring this up, but while testing the changes I noticed a bug, and I wanted to find out if there was a reason why the `model.endpoint` attribute needs to always be from a default env? Currently the [implementation](https://github.com/caraml-dev/merlin/blob/main/python/sdk/merlin/model.py#L734) of the `endpoint` getter looks up all the endpoints of a model version to retrieve the first endpoint that's in the default env. I see that this attribute is used for setting the traffic of the model but I'm not sure why it can't be from a non-default env.

The implication of this is that in order to allow a model already deployed to a non-default env to be redeployed to the same env when the `environment_name` is not specified, I'll be unable to check the endpoint returned by the current `endpoint` getter (since the endpoint of the model deployed to the non-default env wouldn't be returned). Actually this would probably affect the way the entire `deploy` function works when redeploying a model 🤔

Reply:
-> I wanted to find out if there was a reason why the model.endpoint attribute needs to always be from a default env?

Good question. I honestly forget why we did it this way in the first place, but inferring from how it is implemented now, I think because ModelVersion's endpoint() method is being called without the environment name parameter, we expect to return the default endpoint. (Again, why the environment name parameter is missing, I forget 🙏🏻)

As you described, the implication of using the ModelVersion.endpoint() method is leading to non-expected behavior. I think we should create new method here to get the correct previous version endpoint that to be redeployed.

Reply:
Okay, I've created a new helper method `_get_deployed_endpoint` to return an existing running (it will select the _first_ by default)/serving endpoint. This endpoint is the one that gets updated by calling the `deploy` method. This should allow users to redeploy their endpoints in a non-default environment in the same environment.

However, that still doesn't properly address the situation (even if we don't face this now) whereby a model version is already deployed to 2 different environments (1 endpoint in each environment) and if the user doesn't specify any environment name as an argument to the `deploy` method - which endpoint should the SDK update? That said I think this scenario isn't too pressing to address right now, we might perhaps want to align ourselves on how to determine the SDK's behaviour in such a scenario. I'll add a card somewhere so we can perhaps pick this up again in the future.

Reply:
In the end, we've decided to replace `_get_deployed_endpoint` with a similar helper method `_get_endpoint_in_environment` that instead returns an endpoint in the environment specified. This environment will be the one specified as an argument in the `deploy` method, otherwise (or if it's none) it will be the default environment. This implementation will make the `deploy` method work in the same way as the `serve_traffic` method and addresses all the concerns raised above.